File: CPP/ByTier/F/1748_F_Circular_Xor_Reversal.cpp
// Problem: CF 1748 F - Circular Xor Reversal
// https://codeforces.com/contest/1748/problem/F

#include <iostream>
// Flowerbox describing the code purpose, algorithms, time and space complexity
/*
Code Purpose:
This code solves the "Circular Xor Reversal" problem where we need to reverse an
array of powers of 2 using XOR operations on adjacent elements in a circular
manner. The approach uses a sequence of operations that effectively shifts
elements to their correct positions in reverse order.

Algorithms/Techniques:
- The algorithm uses a series of carefully chosen indices to perform XOR
operations.
- It constructs three segments of operations to achieve the reversal:
  1. First segment to prepare the array.
  2. Second segment to move elements to their final positions.
  3. Third segment to clean up and finalize the reversed array.

Time Complexity: O(n^2) - worst case due to nested loops in the operation
generation. Space Complexity: O(n) - for storing the sequence of operations.

*/
int rs[300013], t = 0, n, i;
void o(int l, int r) {
  int l_ = l, r_ = --r;
  while (l <= r) {
    for (i = r - 1; i >= l; i--)
      rs[++t] = i % n; // Store indices in reverse order to shift elements
    for (i = l + 1; i < r; i++)
      rs[++t] = i % n; // Store indices in normal order
    l++, r--;          // Move inward
  }
  for (i = l_; i < (l_ + r_) / 2; i++)
    rs[++t] = i % n; // Additional operations to finalize
}
int main() {
  std::cin >> n;
  o(0, n), o((n + 1) / 2, n / 2 + n), o(0, n); // Call operations in sequence
  for (std::cout << t << '\n', i = 1; i <= t; i++)
    std::cout << rs[i] << ' '; // Output all operations
}

// https://github.com/VaHiX/CodeForces/

File: Python/ByTier/F/1748_F_Circular_Xor_Reversal.py
# Problem: CF 1748 F - Circular Xor Reversal
# https://codeforces.com/contest/1748/problem/F

"""
Algorithm: Circular Xor Reversal
Purpose: Reverse an array where each element is initially 2^i, using XOR operations.
Technique: 
- Use a multi-swap strategy to reverse the array.
- The approach involves:
  1. Performing a series of operations to rotate elements towards their correct positions.
  2. Using a helper function `multiswap` to perform a sequence of swaps and XOR operations.
- Time Complexity: O(n^2) in worst case due to multiple nested loops for generating operations.
- Space Complexity: O(n) for storing the operations.

The algorithm ensures that the XOR operations on indices simulate the reversal process.
"""

n = int(input())
ops = []


def multiswap(i, j):
    # Helper function to generate operations for swapping elements in a range
    assert i < j

    m = j - i
    for end in range(m):
        # Add operations in reverse order for the first part
        ops.extend(range(i + end, i + m - end)[::-1])
        # Add operations in normal order for the second part
        ops.extend(range(i + end + 1, i + m - end))
    # Add final part of operations
    ops.extend(range(i, i + (m - 2) // 2 + 1))


# Perform multiswap operations to reverse the array
multiswap(0, n - 1)
if n % 2:
    # For odd n, perform an additional multiswap
    multiswap((n + 1) // 2 - n, (n - 3) // 2)
else:
    # For even n, perform another multiswap
    multiswap(n // 2 - n, n // 2 - 1)
multiswap(0, n - 1)

# Output the number of operations and the indices used
print(len(ops))
print(*[v % n for v in ops])


# https://github.com/VaHiX/CodeForces/